pkg/intermediate/aggregate.go

```diff
@@ -0,0 +1,218 @@
+	SourceAddress uint32
+	// DestinationAddress is the encoded format of destinationIPv4Address
+	DestinationAddress uint32
+	Protocol           uint8
+	SourcePort         uint16
+	DestinationPort    uint16
+}
+
+func InitAggregationProcess(messageChan chan *entities.Message, workerNum int) (*aggregation, error) {
+	if messageChan == nil {
+		return nil, fmt.Errorf("Cannot create aggregation process without message channel.")
+	} else if workerNum <= 0 {
+		return nil, fmt.Errorf("Worker number cannot be <= 0.")
+	}
+	return &aggregation{
+		make(map[Tuple][]entities.Record),
+		sync.RWMutex{},
+		make(chan chan *entities.Message),
+		messageChan,
+		workerNum,
+		make([]*worker, 0),
+	}, nil
+}
+
+func (a *aggregation) Start() {
+	for i := 0; i < a.workerNum; i++ {
+		w := createWorker(i, a.workerPool, a.AggregateMsgBy5Tuple)
+		w.start()
+		a.workerList = append(a.workerList, w)
+	}
+	for message := range a.messageChan {
+		channel := <-a.workerPool
+		channel <- message
+	}
+}
+
+func (a *aggregation) Stop() {
+	for _, worker := range a.workerList {
+		worker.stop()
+	}
+}
+
+// AggregateMsgBy5Tuple gets 5-tuple info from records in message and stores in cache
+func (a *aggregation) AggregateMsgBy5Tuple(message *entities.Message) error {
+	addOriginalExporterInfo(message)
+	if message.Set.GetSetType() == entities.Template { // skip template records
+		return nil
+	}
+	records := message.Set.GetRecords()
+	a.tupleRecordLock.Lock()
+	defer a.tupleRecordLock.Unlock()
+	for _, record := range records {
+		tuple, err := getTupleFromRecord(record)
+		if err != nil {
+			return err
+		}
+		a.correlateRecords(tuple, record)
+	}
+	return nil
+}
+
+func (a *aggregation) GetTupleRecordMap() map[Tuple][]entities.Record {
+	a.tupleRecordLock.RLock()
+	defer a.tupleRecordLock.RUnlock()
+	return a.tupleRecordMap
+}
+
+// correlateRecords fills records info by correlating incoming and current records
+// only records from source node will be added to tupleRecordMap
+func (a *aggregation) correlateRecords(tuple Tuple, record entities.Record) {
+	if _, exist := a.tupleRecordMap[tuple]; !exist {
+		// if flow key does not exist in the cache, just add record to tupleRecordMap cache
+		a.tupleRecordMap[tuple] = make([]entities.Record, 0)
+		a.tupleRecordMap[tuple] = append(a.tupleRecordMap[tuple], record)
+		return
+	}
+	existingRecords := a.tupleRecordMap[tuple]
+	elements := record.GetInfoElements()
+	for _, existingRecord := range existingRecords {
+		existingElements := existingRecord.GetInfoElements()
+		for i, ieWithValue := range existingElements {
+			if ieWithValue.Value == "" {
+				ieWithValue.Value = elements[i].Value
+			}
+		}
+	}
+	for _, ieWithValue := range elements {
+		if ieWithValue.Element.Name == "sourcePodName" && ieWithValue.Value != "" {
+			// add record to tupleRecordMap when the record is from source node
+			a.tupleRecordMap[tuple] = append(a.tupleRecordMap[tuple], record)
+			return
+		}
+	}
```

Comment:
I did not follow this logic. What happens if we get the record from source node first and destination node later?